public/jira/rollup/percent-complete/percent-complete.ts

```diff
@@ -0,0 +1,401 @@
+    return data;
+  }
+  // if there is hard parent data, use it
+  else if (parentIssueOrRelease?.derivedTiming?.totalDaysOfWork) {
+    data = {
+      completedWorkingDays:
+        parentIssueOrRelease.derivedTiming.completedDaysOfWork,
+      totalWorkingDays: parentIssueOrRelease.derivedTiming.totalDaysOfWork,
+      userSpecifiedValues: true,
+      get remainingWorkingDays() {
+        return this.totalWorkingDays - this.completedWorkingDays;
+      },
+    };
+    // make sure we can build an average from it
+    metadata.totalDaysOfWorkForAverage.push(data.totalWorkingDays);
+    return data;
+  }
+
+  // if there is weak children data, use it, but don't use it for other averages
+  else if (childrenRollups.length) {
+    data = sumChildRollups(childrenRollups);
+    return data;
+  }
+  // if there are no children, add to get the uncertainty
+  else {
+    data = emptyRollup();
+    metadata.needsAverageSet.push(data);
+    return data;
+  }
+}
+
+function sumChildRollups(children: PercentCompleteRollup[]) {
+  const userSpecifiedValues = children.every((d) => d.userSpecifiedValues);
+  const totalDays = children.map((child) => child.totalWorkingDays);
+  const completedDays = children.map((child) => child.completedWorkingDays);
+  return {
+    completedWorkingDays: sum(completedDays),
+    totalWorkingDays: sum(totalDays),
+    userSpecifiedValues: userSpecifiedValues,
+    get remainingWorkingDays() {
+      return this.totalWorkingDays - this.completedWorkingDays;
+    },
+  };
+}
+
+/**
+ * @param { JiraIssue[] } issues
+ * @param { PercentCompleteOptions } options
+ */
+export function percentComplete(derivedWorkIssues: DerivedIssue[]) {
+  return completionRollup(derivedWorkIssues);
+}
+
+function groupIssuesByHierarchyLevel<T extends { hierarchyLevel: number }>(
+  issues: T[]
+): T[][] {
+  const result: T[][] = [];
+  for (const issue of issues) {
+    if (!result[issue.hierarchyLevel]) {
+      result[issue.hierarchyLevel] = [];
+    }
+    result[issue.hierarchyLevel].push(issue);
+  }
+  return result;
+}
+
+/**
+ *
+ * @param {Array<DerivedIssue> issues
+ * @returns {Array<RolledupCompletionIssue>}
+ */
+function toCompletionRollups<CustomFields>(
+  issues: IssueOrRelease<CustomFields>[]
+): RolledupCompletionIssue<CustomFields>[] {
+  return issues.map((issue) => ({
+    ...issue,
+    completionRollup: {
+      totalWorkingDays: 0,
+      completedWorkingDays: 0,
+      remainingWorkingDays: 0,
+    },
+  }));
+}
+
+/**
+ *
+ * @param {Array<DerivedIssue> allIssueData
+ * @param {*} options
+ * @returns {{issues: Array<RolledupCompletionIssue>, hierarchyData: Array<IssueTypeData>}}
+ */
+function completionRollup<CustomFields>(allIssueData: DerivedIssue[]) {
+  const completionRollups = toCompletionRollups(allIssueData);
+
+  const groupedIssueData = groupIssuesByHierarchyLevel(completionRollups);
+
+  // Object.groupBy requires "target: ESNext" in tsconfig
+  const issueKeyToChildren = Object.groupBy(
```

Comment:
Note: Object.groupBy is only available in the ESNext target, so preserving the existing code as much as possible meant changing tsconfig target from ES2020 to ESNext